Shiboken is patched

ViacheslavYartsev opened the pull request:
Signed-off-by: Viacheslav Yartsev <sl_yartsev@mail.ru>

ViacheslavYartsev committed:
Shiboken is patched

Signed-off-by: Viacheslav Yartsev <sl_yartsev@mail.ru>

asergunov commented on shibokenmodule/CMakeLists.txt:
```diff
@@ -9,7 +9,7 @@ ${CMAKE_CURRENT_BINARY_DIR}/shiboken/shiboken_module_wrapper.cpp
 
 add_custom_command(OUTPUT ${sample_SRC}
-COMMAND ${shibokengenerator_BINARY_DIR}/shiboken --project-file=${CMAKE_CURRENT_BINARY_DIR}/shibokenmodule.txt ${GENERATOR_EXTRA_FLAGS}
+COMMAND ${shibokengenerator_BINARY_DIR}/shiboken.exe --project-file=${CMAKE_CURRENT_BINARY_DIR}/shibokenmodule.txt ${GENERATOR_EXTRA_FLAGS}
```
Are you sure it's not working without this change?

asergunov commented on tests/libsample/objecttype.h:
```diff
@@ -70,7 +70,7 @@ class LIBSAMPLE_API ObjectType
     void setObjectName(const Str& name);
 
-    inline unsigned long identifier() const { return reinterpret_cast<unsigned long>(this); }
+    inline unsigned long identifier() const { return reinterpret_cast<uintptr_t>(this); }
```
Looks strange. Because return type is not changed, but conversion is changed.

asergunov commented on generator/shiboken/overloaddata.cpp:
```diff
@@ -882,7 +882,7 @@ QString OverloadData::dumpGraph() const
         s << "}" << endl;
     } else {
-        QString argId = QString("arg_%1").arg((ulong)this);
+        QString argId = QString("arg_%1").arg((uintptr_t)this);
```
`reinterpret_cast` please

asergunov commented on libshiboken/basewrapper.cpp:
```diff
@@ -770,7 +770,7 @@ bool isUserType(PyObject* pyObj)
 {
     assert(Shiboken::Object::checkType(pyObj));
-    return reinterpret_cast<Py_hash_t>(pyObj);
+    return reinterpret_cast<uintptr_t>(pyObj);
```
The same here. Are you sure it will not without this change? Cast to type different than return type.

asergunov replied:
Please check if uintptr_t is 64 bits and if it is 64 bit please XOR low and hight 32 bits

asergunov reviewed (changes requested):
Please update

ViacheslavYartsev committed:
fixing after review, using __SIZEOF_POINTER__ to check 64 bit system
use HI_WORD XOR LO_WORD to calculate hash  for 64 bit case

Signed-off-by: Viacheslav Yartsev <slava@ahead.io>

Commit:
revert shibokenmodule/CMakeLists.txt

Signed-off-by: Anton Sergunov <anton@ahead.io>

asergunov commented on tests/libsample/objecttype.h:
```diff
@@ -70,7 +70,15 @@ class LIBSAMPLE_API ObjectType
 
-    inline unsigned long identifier() const { return reinterpret_cast<unsigned long>(this); }
+    inline unsigned long identifier() const {
+#if __SIZEOF_POINTER__ == 8
```
this one is identifier. Not the hash. So it's bad idea to drop bytes here. It's only two options here

- change return type checking all the usages
- add the static map index `unsigned long` to `uintptr_t` and back.

For me the first one looks better

asergunov commented on libshiboken/basewrapper.cpp:
```diff
@@ -770,7 +770,13 @@ bool isUserType(PyObject* pyObj)
 Py_hash_t hash(PyObject* pyObj)
 {
     assert(Shiboken::Object::checkType(pyObj));
+#if __SIZEOF_POINTER__ == 8
```
that's gcc only @ViacheslavYartsev looks like it's better to have something like 
```c++
unsigned long result = reinterpret_cast<uintptr_t>(this);
for(int i=sizeof(unsigned long); i<sizeof(uintptr_t); i+=sizeof(unsigned long)) {
  result = result ^  static_cast<unsigned long>(reinterpret_cast<uintptr_t>(this) >> i*8);
}
return result;
```
Since cycle is depend only on compile-time known variables it will be optimized out by compiller

ViacheslavYartsev replied:
Thx Anton. The nice solution

asergunov reviewed (changes requested):
Please change the `hash` method.

For the `identifier` please check if it's possible to change return type to uintptr_t.

ViacheslavYartsev committed:
change type of identifier() to uintptr_t
using uintptr_t in ~Collector class
change Py_hash_t hash(PyObject* pyObj) function



Signed-off-by: Viacheslav Yartsev <slava@ahead.io>

ViacheslavYartsev committed:
Merge branch 'submite_the_patch' of https://github.com/AheadIO/Shiboken into submite_the_patch

asergunov commented:
I have
```
shiboken/generator/shiboken/overloaddata.cpp: In member function ‘QString OverloadData::dumpGraph() const’:
shiboken/generator/shiboken/overloaddata.cpp:885:64: error: expected type-specifier before ‘uintptr_t’
         QString argId = QString("arg_%1").arg(reinterpret_cast<uintptr_t>(this));
                                                                ^
shiboken/generator/shiboken/overloaddata.cpp:885:64: error: expected ‘>’ before ‘uintptr_t’
shiboken/generator/shiboken/overloaddata.cpp:885:64: error: expected ‘(’ before ‘uintptr_t’
shiboken/generator/shiboken/overloaddata.cpp:885:64: error: ‘uintptr_t’ was not declared in this scope
```

Fixing

Commit:
includes are added

Signed-off-by: Anton Sergunov <anton@ahead.io>

asergunov commented:
I've added includes but now it can't build tests on Linux

```
shiboken/build/tests/samplebinding/sample/collector_wrapper.cpp: In function ‘PyObject* Sbk_CollectorFunc_items(PyObject*)’:
shiboken/build/tests/samplebinding/sample/collector_wrapper.cpp:106:65: error: conversion from ‘std::list<long unsigned int>’ to non-scalar type ‘std::list<unsigned int>’ requested
             std::list<unsigned int > cppResult = cppSelf->items();
```

it's a wrapper generated by the shiboken but type found by shiboken is wrong. It produces
```c++
            PyThreadState* _save = PyEval_SaveThread(); // Py_BEGIN_ALLOW_THREADS
            std::list<unsigned int > cppResult = cppSelf->items();
            PyEval_RestoreThread(_save); // Py_END_ALLOW_THREADS
            pyResult = Shiboken::Conversions::copyToPython(SbksampleTypeConverters[SBK_SAMPLE_STD_LIST_UNSIGNEDINT_IDX], &cppResult);
        }
```
but `cppSelf->items()` returns `list<uintptr_t>` which are `long unsigned int`

`intptr_t` is defined as
```c++
/* Types for `void *' pointers.  */
#if __WORDSIZE == 64
# ifndef __intptr_t_defined
typedef long int		intptr_t;
#  define __intptr_t_defined
# endif
typedef unsigned long int	uintptr_t;
#else
# ifndef __intptr_t_defined
typedef int			intptr_t;
#  define __intptr_t_defined
# endif
typedef unsigned int		uintptr_t;
#endif
```

Seems the cause of problem is missing `#define __WORDSIZE 64` for the parser

asergunov commented:
`__WORDSZE` is defined as 

```c++
#if defined __x86_64__ && !defined __ILP32__
# define __WORDSIZE	64
#else
# define __WORDSIZE	32
#endif```

means we have to let parser know if these definitions are set while

Commit:
__x86_64__ and __ILP32__ are set as in build environment

Signed-off-by: Anton Sergunov <anton@ahead.io>

Commit:
compile warning fix

asergunov commented:
Shiboken still treat uint_ptr wrong.
